src/modules/avformat/link_avdeinterlace.c

```diff
@@ -0,0 +1,617 @@
+	const AVFilter *buffersrc  = avfilter_get_by_name("buffer");
+	const AVFilter *buffersink = avfilter_get_by_name("buffersink");
+	enum AVPixelFormat in_pixel_fmts[] = { -1, -1 };
+	enum AVPixelFormat out_pixel_fmts[] = { -1, -1 };
+	AVRational timebase = (AVRational){ profile->frame_rate_den, profile->frame_rate_num };
+	AVRational framerate = (AVRational){ profile->frame_rate_num, profile->frame_rate_den };
+	AVFilterContext* prev_ctx = NULL;
+	AVFilterContext* avfilter_ctx = NULL;
+	int ret;
+
+	if ( !pdata->avinframe )
+		pdata->avinframe = av_frame_alloc();
+	if ( !pdata->avoutframe )
+		pdata->avoutframe = av_frame_alloc();
+
+	// Set up formats
+	in_pixel_fmts[0] = mlt_to_av_image_format( pdata->informat );
+	out_pixel_fmts[0] = mlt_to_av_image_format( pdata->outformat );
+
+	// Destroy the current filter graph
+	avfilter_graph_free( &pdata->avfilter_graph );
+
+	// Create the new filter graph
+	pdata->avfilter_graph = avfilter_graph_alloc();
+	if( !pdata->avfilter_graph ) {
+		mlt_log_error( self, "Cannot create filter graph\n" );
+		goto fail;
+	}
+	pdata->avfilter_graph->scale_sws_opts = av_strdup("flags=" MLT_SWS_FLAGS);
+
+	// Initialize the buffer source filter context
+	pdata->avbuffsrc_ctx = avfilter_graph_alloc_filter( pdata->avfilter_graph, buffersrc, "in");
+	if( !pdata->avbuffsrc_ctx ) {
+		mlt_log_error( self, "Cannot create image buffer source\n" );
+		goto fail;
+	}
+	ret = av_opt_set_int( pdata->avbuffsrc_ctx, "width", pdata->inwidth, AV_OPT_SEARCH_CHILDREN );
+	if( ret < 0 ) {
+		mlt_log_error( self, "Cannot set src width %d\n", pdata->inwidth );
+		goto fail;
+	}
+	ret = av_opt_set_int( pdata->avbuffsrc_ctx, "height", pdata->inheight, AV_OPT_SEARCH_CHILDREN );
+	if( ret < 0 ) {
+		mlt_log_error( self, "Cannot set src height %d\n", pdata->inheight );
+		goto fail;
+	}
+	ret = av_opt_set_pixel_fmt( pdata->avbuffsrc_ctx, "pix_fmt", in_pixel_fmts[0], AV_OPT_SEARCH_CHILDREN );
+	if( ret < 0 ) {
+		mlt_log_error( self, "Cannot set src pixel format %d\n", in_pixel_fmts[0] );
+		goto fail;
+	}
+	ret = av_opt_set_q( pdata->avbuffsrc_ctx, "sar", sar, AV_OPT_SEARCH_CHILDREN );
+	if( ret < 0 ) {
+		mlt_log_error( self, "Cannot set src sar %d/%d\n", sar.num, sar.den );
+		goto fail;
+	}
+	ret = av_opt_set_q( pdata->avbuffsrc_ctx, "time_base", timebase, AV_OPT_SEARCH_CHILDREN );
+	if( ret < 0 ) {
+		mlt_log_error( self, "Cannot set src time_base %d/%d\n", timebase.num, timebase.den );
+		goto fail;
+	}
+	ret = av_opt_set_q( pdata->avbuffsrc_ctx, "frame_rate", framerate, AV_OPT_SEARCH_CHILDREN );
+	if( ret < 0 ) {
+		mlt_log_error( self, "Cannot set src frame_rate %d/%d\n", framerate.num, framerate.den );
+		goto fail;
+	}
+	ret = avfilter_init_str( pdata->avbuffsrc_ctx, NULL );
+	if( ret < 0 ) {
+		mlt_log_error( self, "Cannot init buffer source\n" );
+		goto fail;
+	}
+	prev_ctx = pdata->avbuffsrc_ctx;
+
+	// Initialize the deinterlace filter context
+	if ( pdata->method <= mlt_deinterlace_method_onefield )
```

Comment:
Currently implemented: onefield, linearblend, yadif_nospatial, yadif, bwdif. FFMpeg does not have some of the  modes from the xine module: bob, weave, greedy. But it does have some others: cubicblend, etc.

Reply:
The "bw" in "bwdif" is for bob and weave, but I do not know if you can put it into one more or another. You could map both to bwdiff. All unknown should fall to some default, maybe yadif since it is both fast and good. bwdif might be better quality, but how is performance compared?

Reply:
> The "bw" in "bwdif" is for bob and weave, but I do not know if you can put it into one more or another. 

bwdif performs motion adaptive deinterlacing based on yadif with the use of w3fdif and cubic interpolation algorithms. There are not options to simplify it down to a classic "bob" or "weave". Personally, I think that bob and weave have fallen out of relevance since we have much higher performance computers now. It is OK with me if our fastest deinterlace methods are one field and linear blend.

> All unknown should fall to some default, maybe yadif since it is both fast and good.

I prefer to let each implementation define its own default. I chose the best option as default for both:
https://github.com/bmatherly/mlt/blob/deinterlace_link/src/modules/avformat/link_avdeinterlace.c#L256
https://github.com/bmatherly/mlt/blob/deinterlace_link/src/modules/xine/common.c#L99

> bwdif might be better quality, but how is performance compared?

In my past experience, bwdif uses more resources. But in this modern day of computing, I think it is negligible. I'll do some performance comparisons to be sure.

I tried to list the types in the enum in ascending order of difficulty/quality:
https://github.com/bmatherly/mlt/blob/deinterlace_link/src/framework/mlt_types.h#L126
the idea is that if a service does not support a requested mode, it will fail down to the next best mode that it does support
https://github.com/bmatherly/mlt/blob/deinterlace_link/src/modules/xine/common.c#L90
When I do some performance testing, I will be sure to put them in best order.